Fix merging conditions for *_through with STI case

jnraine opened the pull request:
_Original work done by @iantropov in https://github.com/rails/rails/pull/12659._

jnraine commented:
@sgrif I've made changes to address your comments. Please let me know if it is acceptable.

There are several failing activerecord test cases after the merge:

For `bundle exec rake test:mysql`:

```
  1) Failure:
ActiveRecord::ConnectionAdapters::MysqlAdapterTest#test_bad_connection_mysql [/Users/Jordan/code/rails/activerecord/test/cases/adapters/mysql/mysql_adapter_test.rb:15]:
[ActiveRecord::NoDatabaseError] exception expected, not
Class: <Mysql::Error>
Message: <"Access denied for user 'rails'@'localhost' to database 'inexistent_activerecord_unittest'">
---Backtrace---
/Users/Jordan/code/rails/activerecord/lib/active_record/connection_adapters/mysql_adapter.rb:452:in `real_connect'
/Users/Jordan/code/rails/activerecord/lib/active_record/connection_adapters/mysql_adapter.rb:452:in `connect'
/Users/Jordan/code/rails/activerecord/lib/active_record/connection_adapters/mysql_adapter.rb:108:in `initialize'
/Users/Jordan/code/rails/activerecord/lib/active_record/connection_adapters/mysql_adapter.rb:36:in `new'
/Users/Jordan/code/rails/activerecord/lib/active_record/connection_adapters/mysql_adapter.rb:36:in `mysql_connection'
/Users/Jordan/code/rails/activerecord/test/cases/adapters/mysql/mysql_adapter_test.rb:17:in `block in test_bad_connection_mysql'
---------------


  2) Failure:
FinderTest#test_condition_local_time_interpolation_with_default_timezone_utc [/Users/Jordan/code/rails/activerecord/test/cases/finder_test.rb:665]:
--- expected
+++ actual
@@ -1 +1 @@
-#<Topic id: 1, title: "The First Topic", author_name: "David", author_email_address: "david@loudthinking.com", written_on: "2003-07-16 14:28:11", bonus_time: "2000-01-01 14:28:00", last_read: "2004-04-15", content: "Have a nice day", important: nil, approved: false, replies_count: 1, unique_replies_count: 0, parent_id: nil, parent_title: nil, type: nil, group: nil, created_at: "2015-02-04 01:15:11", updated_at: "2015-02-04 01:15:11">
+nil



  3) Failure:
FinderTest#test_condition_utc_time_interpolation_with_default_timezone_local [/Users/Jordan/code/rails/activerecord/test/cases/finder_test.rb:647]:
--- expected
+++ actual
@@ -1 +1 @@
-#<Topic id: 1, title: "The First Topic", author_name: "David", author_email_address: "david@loudthinking.com", written_on: "2003-07-16 14:28:11", bonus_time: "2000-01-01 14:28:00", last_read: "2004-04-15", content: "Have a nice day", important: nil, approved: false, replies_count: 1, unique_replies_count: 0, parent_id: nil, parent_title: nil, type: nil, group: nil, created_at: "2015-02-04 01:15:11", updated_at: "2015-02-04 01:15:11">
+nil



  4) Failure:
FinderTest#test_hash_condition_local_time_interpolation_with_default_timezone_utc [/Users/Jordan/code/rails/activerecord/test/cases/finder_test.rb:674]:
--- expected
+++ actual
@@ -1 +1 @@
-#<Topic id: 1, title: "The First Topic", author_name: "David", author_email_address: "david@loudthinking.com", written_on: "2003-07-16 14:28:11", bonus_time: "2000-01-01 14:28:00", last_read: "2004-04-15", content: "Have a nice day", important: nil, approved: false, replies_count: 1, unique_replies_count: 0, parent_id: nil, parent_title: nil, type: nil, group: nil, created_at: "2015-02-04 01:15:11", updated_at: "2015-02-04 01:15:11">
+nil



  5) Failure:
FinderTest#test_hash_condition_utc_time_interpolation_with_default_timezone_local [/Users/Jordan/code/rails/activerecord/test/cases/finder_test.rb:656]:
--- expected
+++ actual
@@ -1 +1 @@
-#<Topic id: 1, title: "The First Topic", author_name: "David", author_email_address: "david@loudthinking.com", written_on: "2003-07-16 14:28:11", bonus_time: "2000-01-01 14:28:00", last_read: "2004-04-15", content: "Have a nice day", important: nil, approved: false, replies_count: 1, unique_replies_count: 0, parent_id: nil, parent_title: nil, type: nil, group: nil, created_at: "2015-02-04 01:15:11", updated_at: "2015-02-04 01:15:11">
+nil


4445 runs, 12166 assertions, 5 failures, 0 errors, 4 skips
```

For `bundle exec rake test:mysql2`:

```
1) Failure:
MysqlConnectionTest#test_bad_connection [/Users/Jordan/code/rails/activerecord/test/cases/adapters/mysql2/connection_test.rb:22]:
[ActiveRecord::NoDatabaseError] exception expected, not
Class: <Mysql2::Error>
Message: <"Access denied for user 'rails'@'localhost' to database 'inexistent_activerecord_unittest'">
---Backtrace---
/Users/Jordan/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/mysql2-0.3.17/lib/mysql2/client.rb:70:in `connect'
/Users/Jordan/.rbenv/versions/2.1.4/lib/ruby/gems/2.1.0/gems/mysql2-0.3.17/lib/mysql2/client.rb:70:in `initialize'
/Users/Jordan/code/rails/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:18:in `new'
/Users/Jordan/code/rails/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:18:in `mysql2_connection'
/Users/Jordan/code/rails/activerecord/test/cases/adapters/mysql2/connection_test.rb:24:in `block in test_bad_connection'
---------------


  2) Failure:
FinderTest#test_condition_local_time_interpolation_with_default_timezone_utc [/Users/Jordan/code/rails/activerecord/test/cases/finder_test.rb:665]:
--- expected
+++ actual
@@ -1 +1 @@
-#<Topic id: 1, title: "The First Topic", author_name: "David", author_email_address: "david@loudthinking.com", written_on: "2003-07-16 14:28:11", bonus_time: "2000-01-01 14:28:00", last_read: "2004-04-15", content: "Have a nice day", important: nil, approved: false, replies_count: 1, unique_replies_count: 0, parent_id: nil, parent_title: nil, type: nil, group: nil, created_at: "2015-02-04 01:17:07", updated_at: "2015-02-04 01:17:07">
+nil

  3) Failure:
FinderTest#test_condition_utc_time_interpolation_with_default_timezone_local [/Users/Jordan/code/rails/activerecord/test/cases/finder_test.rb:647]:
--- expected
+++ actual
@@ -1 +1 @@
-#<Topic id: 1, title: "The First Topic", author_name: "David", author_email_address: "david@loudthinking.com", written_on: "2003-07-16 14:28:11", bonus_time: "2000-01-01 14:28:00", last_read: "2004-04-15", content: "Have a nice day", important: nil, approved: false, replies_count: 1, unique_replies_count: 0, parent_id: nil, parent_title: nil, type: nil, group: nil, created_at: "2015-02-04 01:17:07", updated_at: "2015-02-04 01:17:07">
+nil
```

jnraine commented:
It would appear to be an issue with my local environment as the same failures happen on master.

sgrif commented:
I'll take a look at this tomorrow. Thanks.

egilburg commented on activerecord/lib/active_record/associations/through_association.rb:
```diff
@@ -14,13 +14,15 @@ module ThroughAssociation #:nodoc:
               relation = relation.merge(reflection_scope)
             end
-
+            
```
Please don't add empty space lines.

jnraine replied:
Good catch.

sgrif commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,12 @@
+    *Ivan Antropov*
+
+*   Make ActiveRecord::Relation#unscope affect relations it is merged in
+    to.
```
This looks like it was unintentionally added?

sgrif commented:
Can you squash these into a single commit? You can make sure the original committer still gets credit by adding a line like `[Your Name & Their Name]`. Can you also add more details about what this change fixes, and why it's necessary? The change on its own is rather opaque. A helpful description of why we need this will help future `git blame` ers.

jnraine committed:
Avoid empty result set when using `has_many :through` with STI models

Fixed bug which caused `has_many :through` associations to return an
empty result. This effected `has_many :through` associations with STI
models that shared the same table.

This was caused by merging the intermediate model's scope
(`PrimaryTask` in the example below) into the target scope, resulting
in a query that will always return nothing.

Example:
    # PrimaryTask and SecondaryTask inherit from Task
    class Person < ActiveRecord::Base
      has_many :primary_tasks
      has_many :secondary_tasks, through: :primary_tasks
    end

    Person.first.secondary_tasks # query contained always-false WHERE
    =>  WHERE `tasks`.`type` IN ('PrimaryTask')
          AND `tasks`.`type` IN ('SecondaryTask')

Instead, don't merge scopes if the reflection's table name matches the
scope's table name.

Fixes: #12474

jnraine commented:
@sgrif OK, changes are squashed. I've rewritten the CHANGELOG.md entry and commit message to include the more detail about the issue.

jnraine commented:
@sgrif is anything else you'd like tweaked on for this? Once you're ready to merge it, let me know and I'll fix the merge conflicts.

sgrif commented:
@jnraine I haven't had a chance to give this a final look yet, I will when I have time. No need to repeatedly ping.

sgrif commented:
Looks like the tests for this are insufficient. Your test case passes without the changes to the code.

jnraine commented:
Shoot! I should have checked that these did what was intended before pulling them from the other PR. I'll re-write them when I have a chance.

sgrif commented:
I'd appreciate confirming the issue even exists anymore. I have a feeling it's simply already been resolved.

jnraine commented:
Will do.

jnraine commented:
I've confirmed that is broken on v4.2.0 and fixed in master. Awesome!

naw commented:
Only a subset of this issue is fixed on master.

1. I believe this works on master only as a side effect to changes in how relation merges work, not as an actual fix to the underlying problem.

2. Neither this PR or master address the underlying problem, which is ThroughAssociations need to be smarter about pulling in default scope and STI type for the reflection chain so that it can account for the fact that the table name for the joined table might be different than its usual name due to the main association using the same table name.  

3. Rather than skipping the condition altogether (as this PR does), the condition needs o be applied against the aliased able name assigned by the AssociationScope.

4. Supposing this _was_ fixed on master, what is the procedure to backport to 4-2-stable?

Here is a gist using the ActiveRecord bug report template with failing tests for 4.2 and failing tests for master:

https://gist.github.com/naw/fa55816dae7805355efd

Thank you

@sgrif

kuwabarahiroshi commented:
@sgrif 
I have encountered the same issue with #12474 on 4.2.3, and confirmed this PR fixes it.
Shouldn't it be backported to 4-2-stable?

BTW, I'd support @naw's suggestion, ideally:

> 3.Rather than skipping the condition altogether (as this PR does), the condition needs to be applied against the aliased table name assigned by the AssociationScope.
